src/cyclefold/circuit.rs

```diff
@@ -0,0 +1,222 @@
+  result: Option<Commitment<E>>,
+  scalar: Vec<Option<bool>>,
+}
+
+impl<E: Engine> CyclefoldCircuitInputs<E> {
+  /// TODO
+  pub fn new(
+    commit_1: Option<Commitment<E>>,
+    commit_2: Option<Commitment<E>>,
+    result: Option<Commitment<E>>,
+    scalar: Vec<Option<bool>>,
+  ) -> Self {
+    Self {
+      commit_1,
+      commit_2,
+      result,
+      scalar,
+    }
+  }
+}
+
+/// TODO: docs
+pub struct CyclefoldCircuit<E: Engine> {
+  inputs: Option<CyclefoldCircuitInputs<E>>,
+}
+
+impl<E: Engine> CyclefoldCircuit<E> {
+  /// TODO: docs
+  pub fn new(inputs: Option<CyclefoldCircuitInputs<E>>) -> Self {
+    Self { inputs }
+  }
+
+  fn alloc_witness<CS: ConstraintSystem<<E as Engine>::Base>>(
+    &self,
+    mut cs: CS,
+  ) -> Result<
+    (
+      AllocatedPoint<E>,
+      AllocatedPoint<E>,
+      AllocatedPoint<E>,
+      Vec<AllocatedBit>,
+    ),
+    SynthesisError,
+  > {
+    let commit_1 = AllocatedPoint::alloc(
+      cs.namespace(|| "allocate C_1"),
+      self
+        .inputs
+        .as_ref()
+        .and_then(|inputs| inputs.commit_1.map(|C_1| C_1.to_coordinates())),
+    )?;
+
+    let commit_2 = AllocatedPoint::alloc(
+      cs.namespace(|| "allocate C_2"),
+      self
+        .inputs
+        .as_ref()
+        .and_then(|inputs| inputs.commit_2.map(|C_2| C_2.to_coordinates())),
+    )?;
+
+    let result = AllocatedPoint::alloc(
+      cs.namespace(|| "allocate C_1 + r * C_2"),
+      self
+        .inputs
+        .as_ref()
+        .and_then(|inputs| inputs.result.map(|result| result.to_coordinates())),
+    )?;
+
+    let scalar = self
+      .inputs
+      .as_ref()
+      .ok_or(SynthesisError::AssignmentMissing)?
+      .scalar
+      .iter()
+      .enumerate()
+      .map(|(idx, b)| AllocatedBit::alloc(cs.namespace(|| format!("scalar bit {}", idx)), *b))
+      .collect::<Result<_, _>>()?;
+
+    Ok((commit_1, commit_2, result, scalar))
+  }
+
+  /// TODO: docs
+  pub fn synthesize<CS: ConstraintSystem<<E as Engine>::Base>>(
+    &self,
+    mut cs: CS,
+  ) -> Result<AllocatedPoint<E>, SynthesisError> {
+    let (C_1, C_2, result, r) = self.alloc_witness(cs.namespace(|| "allocate circuit witness"))?;
+
+    // Calculate C_final
+    let r_C_2 = C_2.scalar_mul(cs.namespace(|| "r * C_2"), &r)?;
+
+    let C_final = C_1.add(cs.namespace(|| "C_1 + r * C_2"), &r_C_2)?;
+
+    let (res_x, res_y, res_inf) = result.get_coordinates();
+
+    // enforce C_final = result
+    cs.enforce(
+      || "C_final_x = res_x",
+      |lc| lc + res_x.get_variable(),
+      |lc| lc + CS::one(),
+      |lc| lc + C_final.x.get_variable(),
+    );
+
+    cs.enforce(
+      || "C_final_y = res_y",
+      |lc| lc + res_y.get_variable(),
+      |lc| lc + CS::one(),
+      |lc| lc + C_final.y.get_variable(),
+    );
+
+    cs.enforce(
+      || "C_final_inf = res_inf",
+      |lc| lc + res_inf.get_variable(),
+      |lc| lc + CS::one(),
+      |lc| lc + C_final.is_infinity.get_variable(),
+    );
```

Comment:
It would simplify confidence in correctness if you implemented `AllocatedPoint::enforce_equal` or similar, then used it throughout. Manually writing out all these constraints introduces many opportunities for small errors or inconsistencies, and at best increases the review/audit burden.

Reply:
Even barring that, you could do similar for the components rather than manually constructing all the constraints here.

Reply:
With the changes that @adr1anh proposes I don't actually need to check equality for `AllocatedPoint`s anymore.